cudaPackages: patch for issue #434096

gammieduncan opened the pull request:
## Summary of the Issue

This is a bug ticket associated with this issue: https://github.com/NixOS/nixpkgs/issues/434096

The bug was that running with NIX_DEBUG set to a level greater than zero (both with integer and string literals) caused RUNPATH entries to be missing from the produced binaries. In the original issue, we saw that when building saxpy:

`NIXPKGS_CONFIG=~/.config/nixpkgs/config-sm_89.nix nom build --impure .#cudaPackages_12.saxpy`

where

`
$ cat ~/.config/nixpkgs/config-sm_89.nix
{
  allowUnfree = true;
  cudaSupport = true;
  cudaCapabilities = [ "8.9" ];
}
`

we got:

`$ patchelf --print-rpath ./result/bin/saxpy 
/run/opengl-driver/lib:/nix/store/bpgbzpzv10raqcvk55xsrhb9vbg96r6g-libcublas-12.8.4.1-lib/lib:/nix/store/31wgl4db2zq67jlw6nk1rzggb36isc44-cuda_cudart-12.8.90-lib/lib:/nix/store/lmn7lwydprqibdkghw7wgcn21yhllz13-glibc-2.40-66/lib:/nix/store/fkw48vh7ivlvlmhp4j30hy2gvg00jgin-gcc-14.3.0-lib/lib`

but when we set `NIX_DEBUG=1`, we get:

```
$ patchelf --print-rpath ./result/bin/saxpy
/run/opengl-driver/lib:/nix/store/lmn7lwydprqibdkghw7wgcn21yhllz13-glibc-2.40-66/lib:/nix/store/fkw48vh7ivlvlmhp4j30hy2gvg00jgin-gcc-14.3.0-lib/lib
```

i.e. our cuda packages are dropped.

## Root Cause Analysis
  
This issue happens during CMake’s install step, not the link step. Our setup hooks is stuffing a semicolon-separated list into cmakeFlags as a plain string, and later the Nix CMake builder expands cmake $cmakeFlags without quotes. With NIX_DEBUG=1, xtrace (set -x) makes this expansion visible and the embedded `;` acts as a shell command separator, so the rest of your cmake command becomes a new shell command. That can drop or reset cache entries/flags (including your CMAKE_INSTALL_RPATH*) and you end up with CMake recomputing/overwriting the install RPATH.

The culprit is here:

```
addToSearchPathWithCustomDelimiter ";" CUDAToolkit_ROOT "$path"
addToSearchPathWithCustomDelimiter ";" CUDAToolkit_INCLUDE_DIR "$path/include"
appendToVar cmakeFlags "-DCUDAToolkit_INCLUDE_DIR=$CUDAToolkit_INCLUDE_DIR"
appendToVar cmakeFlags "-DCUDAToolkit_ROOT=$CUDAToolkit_ROOT"
```

$CUDAToolkit_ROOT / $CUDAToolkit_INCLUDE_DIR often look like "/nix/store/...;/nix/store/...". If cmakeFlags is later expanded unquoted, those `;`s split the command.

This PR patches the setup hook to use cmakeFlagsArray with guards, avoiding unquoted semicolons. The change was just to replace string appends with array-safe flags for semicolon-joined values.

## Things done

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [X] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gammieduncan committed:
cudaPackages: patch for issue#434096

gammieduncan commented:
> None of special `*RPATH` variables should be necessary? There must be a bug in set up hooks. Sorry I didn't acquaint with the posts in the issue yet, tbd
> […](#)
> On 12 September 2025 21:26:16 UTC, "nix-owners[bot]" ***@***.***> wrote: @nix-owners[bot] requested your review on: [NixOS/nixpkgs#442479](https://github.com/NixOS/nixpkgs/pull/442479) cudaPackages: patch for issue#434096.

So my understanding of what's happening is:

- the drop we see now happens at CMake install time, not link time. With NIX_DEBUG≥1, ld-wrapper.sh prints debug to stderr; CMake parses stderr for implicit link dirs/libs, misclassifies “toolchain vs non‑toolchain” paths, then generates file(RPATH_CHANGE … NEW_RPATH ""), stripping CUDA.
- Nix’s fixup/shrink-rpath only removes superfluous entries; it does not re-add what CMake has already erased. The current setup hooks don’t restore CUDA libs either (we only add /run/opengl-driver via autoAddDriverRunpath).
- The cmake setup hook in nixpkgs does not set CMAKE_INSTALL_RPATH* globally, and lots of packages set these flags themselves when needed.

I could drop the two “skip” flags (i.e. CMAKE_SKIP_INSTALL_RPATH=false and CMAKE_SKIP_RPATH=false), but keep the two that prevent the strip and make CUDA explicit (i.e. CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON CMAKE_INSTALL_RPATH=${lib.makeLibraryPath [ (getLib libcublas) cuda_cudart ]}.

Alternately, the CUDA setup hook could append -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=ON (and possibly extend CMAKE_INSTALL_RPATH using known CUDA components). 

This is my first PR in this repo, so if you think I'm coming at this from the wrong angle I trust your judgement. Is my breakdown of the issue incorrect?

gammieduncan committed:
cudaPackages: patch for issue#434096 modifying setup hook

SomeoneSerge commented:
Oh wow, impressive. Cursed.

Did you figure out, in...

> Nix CMake builder expands cmake `$cmakeFlags`

..., what does "cmake builder" refer to? At least in `cmake/setup-hook.sh` we sure do do the quoting. I'm suspecting we must be abusing some kind of a subshell, but even then xstrace should go into stderr?

SomeoneSerge commented:
> But then the problem happens in callers that bypass this builder and invoke cmake $cmakeFlags unquoted (some external packages do this, like I saw it in pkgs/top-level/emscripten-packages.nix): raw ; in list-valued -D…= splits the shell command.

Ah! I don't think[^1] expansions or semicolons inside unquoted `$cmakeFlags` there are interpreted, but this immediately reminds me that there's a few places in `python-packages/` where we (or at least I) abuse `os.environ.get("cmakeFlags", "").split()` in `setup.py` and the like. Notably, this abuse won't work with `cmakeFlagsArray`... This `split()` only "interprets" whitespaces, not semicolons, so it's implicitly "ok" for `cmakeFlags`, but it's far from obvious

Leads to two conclusions:

1. Work required to make python-packages and the like support `*Array` variables which are not exported in the environment.
2. We should rather fix those "abusive callers" than avoid passing them the inputs by moving them to `*Array` variables?

[^1]: ```bash
      ❯ flags="; echo foo"
      ❯ echo $flags
      ; echo foo
      ```

gammieduncan commented:
> > But then the problem happens in callers that bypass this builder and invoke cmake $cmakeFlags unquoted (some external packages do this, like I saw it in pkgs/top-level/emscripten-packages.nix): raw ; in list-valued -D…= splits the shell command.
> 
> Ah! I don't think[1](#user-content-fn-1-e6e9869b5abc7b7c3d4d1b1bc2e1d326) expansions or semicolons inside unquoted `$cmakeFlags` there are interpreted, but this immediately reminds me that there's a few places in `python-packages/` where we (or at least I) abuse `os.environ.get("cmakeFlags", "").split()` in `setup.py` and the like. Notably, this abuse won't work with `cmakeFlagsArray`... This `split()` only "interprets" whitespaces, not semicolons, so it's implicitly "ok" for `cmakeFlags`, but it's far from obvious
> 
> Leads to two conclusions:
> 
> 1. Work required to make python-packages and the like support `*Array` variables which are not exported in the environment.
> 2. We should rather fix those "abusive callers" than avoid passing them the inputs by moving them to `*Array` variables?
> 
> ## Footnotes
> 1. ```
>      ❯ flags="; echo foo"
>      ❯ echo $flags
>      ; echo foo
>    ```
>    
>    
>    [↩](#user-content-fnref-1-e6e9869b5abc7b7c3d4d1b1bc2e1d326)

Thanks, unless you think this should be separated into a separate PR/issue, I just pushed a targeted sweep to address the abusive callers (still keeping back-compat):

- Python hook: export CMAKE_FLAGS_NL (newline-delimited args) when cmakeFlagsArray is present (`pkgs/development/interpreters/python/setup-hook.sh`).
- python-packages: stop doing .split() on cmakeFlags; prefer CMAKE_FLAGS_NL, fall back to `shlex.split(os.environ["cmakeFlags"])`. Updated dm-tree and the vLLM patch (0002-setup.py-nix-support-respect cmakeFlags.patch) accordingly.
- Shell caller: fix `pkgs/top-level/emscripten-packages.nix` to pass CMake flags via a quoted array (`concatTo …; emcmake cmake … "${flagsArray[@]}"`).

This should let existing cmakeFlags users to keep working, while also allowing list-valued -D… definitions to live safely in the array path.

ConnorBaker commented:
Thanks for working on this @gammieduncan! I'll try to take a look at this on the weekend.

gammieduncan commented:
> > But then the problem happens in callers that bypass this builder and invoke cmake $cmakeFlags unquoted (some external packages do this, like I saw it in pkgs/top-level/emscripten-packages.nix): raw ; in list-valued -D…= splits the shell command.
> 
> Ah! I don't think[1](#user-content-fn-1-d7eef65fc4c3671d9f50a0f3fbd13aaa) expansions or semicolons inside unquoted `$cmakeFlags` there are interpreted, but this immediately reminds me that there's a few places in `python-packages/` where we (or at least I) abuse `os.environ.get("cmakeFlags", "").split()` in `setup.py` and the like. Notably, this abuse won't work with `cmakeFlagsArray`... This `split()` only "interprets" whitespaces, not semicolons, so it's implicitly "ok" for `cmakeFlags`, but it's far from obvious
> 
> Leads to two conclusions:
> 
> 1. Work required to make python-packages and the like support `*Array` variables which are not exported in the environment.
> 2. We should rather fix those "abusive callers" than avoid passing them the inputs by moving them to `*Array` variables?
> 
> ## Footnotes
> 1. ```
>      ❯ flags="; echo foo"
>      ❯ echo $flags
>      ; echo foo
>    ```
>    
>    
>    [↩](#user-content-fnref-1-d7eef65fc4c3671d9f50a0f3fbd13aaa)

Thanks, unless you think this should be separated into a separate PR/issue, I just pushed a targeted sweep to address the abusive callers (still keeping back-compat):

* Python hook: export CMAKE_FLAGS_NL (newline-delimited args) when cmakeFlagsArray is present (pkgs/development/interpreters/python/setup-hook.sh).
* python-packages: stop doing .split() on cmakeFlags; prefer CMAKE_FLAGS_NL, fall back to shlex.split(os.environ["cmakeFlags"]). Updated dm-tree and the vLLM patch (0002-setup.py-nix-support-respect cmakeFlags.patch) accordingly.
* Shell caller: fix pkgs/top-level/emscripten-packages.nix to pass CMake flags via a quoted array (concatTo …; emcmake cmake … "${flagsArray[@]}").

This should let existing cmakeFlags users to keep working, while also allowing list-valued -D… definitions to live safely in the array path. 

@ConnorBaker thanks for taking a look - had an onsite at work this week, otherwise would've gotten to this sooner. The scope of this issue has expanded a bit, happy to break it up if need be

gammieduncan commented:
Looks like these two jobs timed out after 60 minutes:

- PR / Build / aarch64-linux: shell, docs (pull_request_target)
- PR / Build / x86_64-linux: shell, docs, lib, tarball (pull_request_target)

It could be because I made changes to the python setup hook, so it's rebuilding a bunch of things, I'm not sure if that's much longer than expected though. I don't see any errors in the build. We could rerun, but I'd also like input on if the changes were a) necessary for this issue, or b) if the build is taking longer than expected.

SomeoneSerge commented:
So how do we go about this?

We still need to locate the piece of code that you say interprets `;` as a shell `;`, because that's indubitably a bug, and it seems local, and fixing it will already treat the symptom. The "list-valued `-D`s" already are safe-enough in `cmakeFlags`, as long as we don't `sh -c "... ${cmakeFlags}"`

Then, each of those monkey-patched `setup.py` scripts is also bogus, because it either only reads the environment-exported `cmakeFlags`, or the JSON-exported `cmakeFlags`, but never the dynamically populated bash list. Devising a local solution for this would also help with the future (ongoing? let's pretend and call it "ongoing") migration to `__structuredAttrs`. Without changing much, and without touching the python or cmake hooks that are deep in the bootstrap chain, you could `makeSetupHook ...` anew, and prepare a bash function that would move things from `cmakeFlagsArray` to somewhere an external process can read (the environment, or the JSON). Each time you add a word or a line to the environment, you'd also remove it from the bash list, to ensure there's no duplication. Then we can have those specific setup-py-monkey-patching packages import your hook, and run the helper function sometime before the configuration phase. That's all still stateful and somewhat ugly, but hopefully a step towards untying our hands.

mweinelt commented:
Opposed to changing the fundamental python setup-hook without further reasoning.

gammieduncan commented:
Also @ConnorBaker I noticed you mentioned a different variant of this issue in your cudaPackages repo: https://github.com/ConnorBaker/cuda-packages where you say:

> Discovered in the process of examining saxpy's build, using declare NIX_DEBUG=4 and export NIX_DEBUG=4 yield different logs and results!
Only export shows the before and after flags used with toolchain invocations -- declare does not!
Only declare yields an output with the same RUNPATH as the original -- with export, entries are missing!

and I wanted to let you know that that issue is also resolved now, for the same reasons as I provided in this PR description. I verified that was the case with this bash script:

```
export NIXPKGS_ALLOW_UNFREE=1
common_flags="--impure --no-substitute -L .#cudaPackages_12.saxpy"

rm -f result
nix build $common_flags |& tee base.log
cp -L result/bin/saxpy saxpy.base

rm -f result
declare NIX_DEBUG=4
nix build $common_flags |& tee decl4.log
cp -L result/bin/saxpy saxpy.decl4

rm -f result
export NIX_DEBUG=4
nix build $common_flags |& tee exp4.log
cp -L result/bin/saxpy saxpy.exp4

echo "== RUNPATHs (should all be identical if fixed) =="
for f in base decl4 exp4; do
  echo "--- saxpy.$f"
  patchelf --print-rpath saxpy.$f
done

echo "== DIFF RUNPATH =="
diff -u <(patchelf --print-rpath saxpy.base)  <(patchelf --print-rpath saxpy.decl4) || true
diff -u <(patchelf --print-rpath saxpy.base)  <(patchelf --print-rpath saxpy.exp4)  || true

echo "== HASHES (should match if fixed) =="
sha256sum saxpy.base saxpy.decl4 saxpy.exp4
echo "== DIFF HASHES =="
diff -u <(sha256sum saxpy.base) <(sha256sum saxpy.decl4) || true
diff -u <(sha256sum saxpy.base) <(sha256sum saxpy.exp4)  || true

echo "== LOGS show extra toolchain lines only when exported (expected) =="
echo "--- base (no NIX_DEBUG) ---"
grep -Ei 'cc|ld|flags|invocation' base.log | head -n 20 || true
echo "--- decl4 (declare only) ---"
grep -Ei 'cc|ld|flags|invocation' decl4.log | head -n 20 || true
echo "--- exp4 (exported) ---"
grep -Ei 'cc|ld|flags|invocation' exp4.log | head -n 50 || true

echo "== Cache check (should see compilation, not 'copying path from https://cache.nixos.org') =="
grep -E 'copying path.*cache.nixos.org' base.log decl4.log exp4.log || echo "OK: no cache hits seen"
```

and got the following results:

> == RUNPATHs (should all be identical if fixed) ==
--- saxpy.base
/run/opengl-driver/lib:/nix/store/cwy7010iwla9b2v1fx82sp66v12r913x-libcublas-12.8.4.1-lib/lib:/nix/store/xccbzbpcn8r506zdvhvbkqkilhlrh3c5-cuda_cudart-12.8.90-lib/lib:/nix/store/8p33is69mjdw3bi1wmi8v2zpsxir8nwd-glibc-2.40-66/lib:/nix/store/ysdkxvcvy2sy36sqigkyqanixm76z2xh-gcc-14.3.0-lib/lib
--- saxpy.decl4
/run/opengl-driver/lib:/nix/store/cwy7010iwla9b2v1fx82sp66v12r913x-libcublas-12.8.4.1-lib/lib:/nix/store/xccbzbpcn8r506zdvhvbkqkilhlrh3c5-cuda_cudart-12.8.90-lib/lib:/nix/store/8p33is69mjdw3bi1wmi8v2zpsxir8nwd-glibc-2.40-66/lib:/nix/store/ysdkxvcvy2sy36sqigkyqanixm76z2xh-gcc-14.3.0-lib/lib
--- saxpy.exp4
/run/opengl-driver/lib:/nix/store/cwy7010iwla9b2v1fx82sp66v12r913x-libcublas-12.8.4.1-lib/lib:/nix/store/xccbzbpcn8r506zdvhvbkqkilhlrh3c5-cuda_cudart-12.8.90-lib/lib:/nix/store/8p33is69mjdw3bi1wmi8v2zpsxir8nwd-glibc-2.40-66/lib:/nix/store/ysdkxvcvy2sy36sqigkyqanixm76z2xh-gcc-14.3.0-lib/lib
== DIFF RUNPATH ==
== HASHES (should match if fixed) ==
8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.base
8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.decl4
8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.exp4
== DIFF HASHES ==
--- /dev/fd/63	2025-09-23 21:37:31.196386100 -0400
+++ /dev/fd/62	2025-09-23 21:37:31.204386295 -0400
@@ -1 +1 @@
-8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.base
+8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.decl4
--- /dev/fd/63	2025-09-23 21:37:31.238387123 -0400
+++ /dev/fd/62	2025-09-23 21:37:31.223386758 -0400
@@ -1 +1 @@
-8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.base
+8b0beb33cb287d6491295e9ca8d818e8279c88455c367bd82aefd4ef45219d3f  saxpy.exp4
== LOGS show extra toolchain lines only when exported (expected) ==
--- base (no NIX_DEBUG) ---
--- decl4 (declare only) ---
--- exp4 (exported) ---
== Cache check (should see compilation, not 'copying path from https://cache.nixos.org') ==
OK: no cache hits seen

which shows that `declare NIX_DEBUG=4`, `export NIX_DEBUG=4`, and `export NIX_DEBUG=1` all produce identical RUNPATHs (patchelf --print-rpath …) and identical SHA256s of the binary. The logs differ slightly, but in an expected way: Extra debug output appears only when NIX_DEBUG is exported (as expected, because only exported vars reach child processes and builders).

(Hopefully this isn't too verbose, just noticed this in your repo and wanted to communicate that it was no longer an issue)

SomeoneSerge commented:
I didn't test but AFAICT we've merged this PR in the state where

- we unbreak `NIX_DEBUG=4` derivations,
- at the cost of breaking python packages that were not broken before?

SomeoneSerge commented:
> CMake/FindCUDAToolkit can discover CUDA via standard search (CMAKE_PREFIX_PATH from buildInputs) 

IIRC no they can only be discovered via `CUDAToolkit_ROOT`. However, I think already in https://github.com/NixOS/nixpkgs/pull/233581/files I had this (mis)conception that the usage of `CUDAToolkit_ROOT` as an environment variable was deprecated in favour of the CMake `-D` option, and looking at the docs right now that doesn't seem to be the case. Maybe we should completely delete the env hook extending `cmakeFlags{,Array}`?
